tsconfig.json

```diff
@@ -6,10 +6,10 @@
       "path": "./.github/workflows/scripts/tsconfig.json"
     },
     {
-      "path": "./packages/govuk-frontend-review/tsconfig.json"
+      "path": "./packages/govuk-frontend-review/tsconfig.dev.json"
```

Comment:
Going through commit by commit, so this may be explained later, but does the `.dev` in the filename allow TypeScript to do anything automagically, or is it only for flagging its purpose to us?

Reply:
Nothing magic, I'll explain

Editors like Visual Studio Code "find the nearest tsconfig.json" so they missed out checks in `tsconfig.build.json`

Consider these two commits optional, but I'll explain further:

* **Move “development only” type checks to `tsconfig.dev.json`**
* **Ensure “build only” type checks also appear in editors**

The comment above shows us move `tsconfig.json` → `tsconfig.dev.json`

But later recreates a new wrapper `tsconfig.json` to enable checks across both:

```json
{
  "extends": "../../tsconfig.base.json",
  "files": [],
  "references": [
    {
      "path": "./tsconfig.build.json"
    },
    {
      "path": "./tsconfig.dev.json"
    }
  ]
}
```

We'd use each config as follows:

### tsconfig.dev.json
* Development types like `"jest", "jest-puppeteer"` so globals like `page` and `describe()` get types
* Type checks tasks, tests and other build tools

### tsconfig.build.json
* No development types like `"jest", "jest-puppeteer"` so we can flag accidental globals outside tests
* Optionally locks in `rootDir` and `outDir` to ensure we're not including development packages
* Optionally enables `noEmit: false` to build types, so checks verify [module](https://www.typescriptlang.org/tsconfig#module) + [target](https://www.typescriptlang.org/tsconfig#target) flavours
* Optionally enables strict mode which we wouldn't necessarily use in tests
* Excludes `"**/*.test.*"` files

But see all the issues "live" in the editor, not just in GitHub Actions workflows

Reply:
Oh and I've split out type checks into:

1. `npm run lint:types`
2. `npm run build:types`

So we see "build only" checks during the build step, but test + development things in the linting step

I've added a spike here on what publishing types might look like:

* https://github.com/alphagov/govuk-frontend/pull/4094

You quite quickly forget what it's like without autocomplete, types, descriptions in other projects

Reply:
All good changes - thanks for thinking this stuff through.